### DISCUSSION TITLE
How to author an out-of-process WinRT exe server with CsWinRT?

### DISCUSSION THREAD

**Discussion by hez2010** [microsoft/CsWinRT#1064] [Q&A] [answered]
I found this sample: https://github.com/microsoft/DesktopBridgeToUWP-Samples/tree/master/Samples/WinformsOutOfProcessWinRTComponent

I'm trying to replace the C++ WinRT server with C# using CsWinRT:

WinUIApp1.Server:
```csharp
using System.Runtime.InteropServices.WindowsRuntime;
using Windows.Foundation;
using Windows.Storage;
using Windows.Storage.Streams;

namespace WinUIApp1.Server;

class Program
{
    static async Task Main()
    {
        while (true)
        {
            await Task.Delay(1000);
        }
    }
}

public sealed class TestClass
{
    public IAsyncOperation<IInputStream> GetFileStreamAsync(string fileName)
    {
        Console.WriteLine("Received IPC Call");
        return AsyncInfo.Run(async _ =>
        {
            var file = await StorageFile.GetFileFromPathAsync(fileName);
            return await file.OpenSequentialReadAsync();
        });
    }
}
```

and the csproj is:
```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net6.0-windows10.0.19041.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
    <CsWinRTComponent>true</CsWinRTComponent>
    <Platforms>x86;x64;arm64</Platforms>
    <UseWindowsForms>true</UseWindowsForms>
    <SelfContained>true</SelfContained>
    <RuntimeIdentifier>win10-x64</RuntimeIdentifier>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.Windows.CsWinRT" Version="1.4.1" />
  </ItemGroup>

</Project>
```

Then in my client app I referenced the generated `WinUIApp.Server.winmd` and wrote below code:
```csharp
var server = new TestClass();
```

And finally I created a package and registered the `WinUIApp.Server.TestClass` as out-of-process activatiableClass:

```xml
    <Extension Category="windows.activatableClass.outOfProcessServer">
      <OutOfProcessServer ServerName="WinUIApp1.Server" uap5:IdentityType="activateAsPackage" uap5:RunFullTrust="true">
        <Path>WinUIApp1.Server\WinUIApp1.Server.exe</Path>
        <Instancing>singleInstance</Instancing>
        <ActivatableClass ActivatableClassId="WinUIApp1.Server.TestClass"></ActivatableClass>
      </OutOfProcessServer>
    </Extension>
```

But it doesn't work. While calling the `new TestClass()`, I can see the server exe being launched successfully, but then the client stuck forever.

A minimal repro:
[WinUIApp1.zip](https://github.com/microsoft/CsWinRT/files/7701461/WinUIApp1.zip)

**Comment by angelazhangmsft** [comment]
@hez2010, we currently don't have a sample for OOP WinRT, but there is a sample on OOP COM: https://github.com/dotnet/samples/tree/main/core/extensions/OutOfProcCOM

For WinRT, you would use the same model as the OutofProcCOM sample but replace CoRegisterClassObject with [RoRegisterActivationFactories](https://docs.microsoft.com/windows/win32/api/roapi/nf-roapi-roregisteractivationfactories), and use [RoRevokeActivationFactories](https://docs.microsoft.com/windows/win32/api/roapi/nf-roapi-rorevokeactivationfactories) on shutdown to unregister the activation factories.

**Reply by hez2010** [reply]
Thank you. I will have a try.

**Reply by hez2010** [reply]
Is there any sample about usage to [RoRegisterActivationFactories](https://docs.microsoft.com/windows/win32/api/roapi/nf-roapi-roregisteractivationfactories) in C#?

**Reply by angelazhangmsft** [reply]
@hez2010 we don't have a sample today, but I've opened #1177 and it's on our backlog to look at.

**Reply by hez** [reply]
👋 Dunno anything about this, sorry! Let me know if you want some consultation.

**Comment by cay7man** [comment]
is there any update? Any sample available now?

**Reply by hez2010** [reply, marked as answer]
I made a sample for this: https://github.com/hez2010/WinRTServer

Note here we need two workarounds:

1. The winmd needs to be removed from `_AppxWinmdFilesToHarvest`, see the target `RemoveOutOfProcWinMD` in the wapproj. Otherwise an incorrect `inProcessServer` entry will be added to the manifest despite there's no `inProcessServer` at all, which will make the build fail.
2. The server project also needs to include the winmd in the `ItemGroup`, otherwise the server implementation (`WinRTServer.dll`) will be removed from the package layout if you publish it to appx package, which can lead to failure while starting the server.

**Reply by wherewhere** [reply]
Dose it supports UWP?

**Reply by hez2010** [reply]
Yes.

**Reply by wherewhere** [reply]
Wow!

**Comment by wherewhere** [comment]
[Dev Home](https://github.com/microsoft/devhome/blob/main/extensionsdk%2FMicrosoft.Windows.DevHome.SDK.Lib%2FExtensionServer.cs) is using oop to create extension by cs/winrt, and [winget](https://github.com/microsoft/winget-cli/tree/master/src%2FWinGetServer) oop is written in cpp/winrt. I make a [Loopback Exemption Manager](https://github.com/wherewhere/Windows-Loopback-Exemption-Manager) in UWP by oop. But I don't know how to use wrl to manage server. So I use the destructor to watching disposing...
